src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionAnnotatedTypeFactory.java

```diff
@@ -220,14 +236,31 @@ protected void postAnalyze(ControlFlowGraph cfg) {
               });
     }
 
-    System.out.println("Known Bindings:");
-    DependencyInjectionAnnotatedTypeFactory.knownBindings.forEach(
-        (key, value) -> {
-          System.out.print(String.format("<Key: %s, Value: %s>\n", key, value));
-        });
+    super.postAnalyze(cfg);
+  }
 
-    System.out.println();
+  @Override
+  public TreeAnnotator createTreeAnnotator() {
+    return new ListTreeAnnotator(
+        new DependencyInjectionTreeAnnotator(this), super.createTreeAnnotator());
+  }
 
-    super.postAnalyze(cfg);
+  public class DependencyInjectionTreeAnnotator extends TreeAnnotator {
+
+    public DependencyInjectionTreeAnnotator(AnnotatedTypeFactory annotatedTypeFactory) {
+      super(annotatedTypeFactory);
+    }
+
+    @Override
+    public Void visitMethod(MethodTree tree, AnnotatedTypeMirror p) {
+      ExecutableElement element = TreeUtils.elementFromDeclaration(tree);
+      if (ElementUtils.hasAnnotation(element, Provides.class.getName())) {
+        // TODO: Put fully qualified names of classes in knownBindings
+        knownBindings.put(tree.getReturnType().toString(), p.getUnderlyingType().toString());
```

Comment:
`TypeMirror#toString`'s specification (https://docs.oracle.com/en/java/javase/17/docs/api/java.compiler/javax/lang/model/type/TypeMirror.html#toString()) says that "any names embedded in the result are qualified if possible", so this should be fine as-is. I do think you need to add some tests to make sure that that is true in practice.

Reply:
I'll look into this